### ISSUE TITLE
Clarify what is meant by "When user interface components receive keyboard focus," in SC 2.4.11

### ISSUE THREAD

**Issue by mraccess77** [w3c/wcag#2219]
I had understood this to meant that when the user scrolled the component out of view that it didn't need to have a focus indicator - but that while it is visible and focus hasn't been moved the focus needed to persist.   Others bring up the point that it only means at the exact moment it receives focus does it need to have an indicator and the indicator can disappear.    Can we please define what is required.   In addition, if is in fact that it only has to briefly appear on focus and can disappear a second later - are there any other success criteria in WCAG that do require focus persist while the control is visible in the viewport and the user has not moved focus?

This difference seems to be related to discussion in #2201

**Comment by patrickhlauke** [comment]
I'm sure we discussed some of this previously in some other issue, but can't find it now. I think it was in the context of whether something like `:focus-visible` would also be acceptable, and whether or not controls need to have persistent visible focus indication even if they were focused programmatically or from a mouse click/touch tap...

**Comment by stevefaulkner** [comment]
I find the idea that a transitory indication of focus is adequate to conform both incorrect and objectionable. Focus indicators represent a state of being focused, change in the display indication when the control is in a focused state implies a change in state, which has not occurred.

**Comment by awkawk** [comment]
I've had questions about this as well. @mbgower had some examples I think.

I've grappled with this SC a lot - just as we are asking what the examples are where the focus needs to disappear, I'm also wondering how much of a problem it is that the focus is not persistent. I can see how it is a problem if the focus isn't shown - even with a mouse people have this location problem and Windows has included a feature to help people find the mouse (https://support.microsoft.com/en-us/windows/find-your-mouse-pointer-fast-dbc1d222-778c-da15-5218-cb8336074554). That said, I know that when I use the keyboard to navigate that I regularly lose track of the focus even when it is quite prominent - I'm thinking about other things / the UI has a lot going on and even though the focus is prominent it can still get lost / etc - and the way that I reorient is to tab a couple of times and shift-tab if needed.

Has focus disappearance been a major problem that 2.4.7 Focus Visible hasn't been addressing?

**Comment by mraccess77** [comment]
I would consider a fading indicator a failure of SC 2.4.7.

**Comment by bruce-usab** [comment]
I must confess, it had not occurred to me to consider a fading indicator under SC 2.4.7.  Sorry to say, I do not think a fail is unambiguous.  Ping to my colleague @kengdoj to say if this detail is covered under Trusted Tester.

I will also note that, prompted by [this 2.4.11 thread](https://github.com/w3c/wcag/issues/2201) I am using the Chrome enhanced focus.  The feature has to make interesting choices when focus moves to anchors which are not keyboard operable.

**Comment by awkawk** [comment]
I have to agree with Bruce. I don't think that 2.4.7 makes it clear that the focus must be persistently visible.

**Comment by stevefaulkner** [comment]
@bruce-usab the understanding doc is unambiguous:


> The focus indicator must not be time limited, when the keyboard focus is shown it must remain.

source: https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html

**Comment by stevefaulkner** [comment]
If we start down the path of saying that if it does not explicitly state that it must be persistent then many a success criterion could also be open to that interpretation. 

> Success Criterion 2.4.6 Headings and Labels
Headings and [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) describe topic or purpose. 

_but only for the first 20 seconds of display…_

**Comment by alastc** [comment]
Indeed, we've had that conversation in the context of labels before, and we came down on things needing to be persistent unless stated otherwise.

**Comment by bruce-usab** [comment]
I appreciate the notes that persistence is included in understanding.  Any ideas where we might include that expectation somewhere in the main document?  I would not like to see it added on an SC-by-SC basis.

But here is an unhappy thought:  If persistence is added as a condition to 2.4.11, there is an implication that persistence is _not_ required by 2.4.7.

**Comment by patrickhlauke** [comment]
> But here is an unhappy thought: If persistence is added as a condition to 2.4.11, there is an implication that persistence is not required by 2.4.7.

At least the normative wording for 2.4.7 does not include any of the problematic "when it receives focus" language, which caused this whole discussion in the first place here for 2.4.11.

2.4.7 is problematic anyway because it didn't normatively define what "visible" actually means, so I think the ship has already sailed in terms of unintended readings of 2.4.7.

**Comment by awkawk** [comment]
I've been thinking about whether the focus appearance SC should basically copy 2.4.7 and add in the new bit with a definition for "highly visible keyboard focus indicator".

Like this:
Any keyboard operable user interface has a mode of operation where the keyboard focus is displayed using a highly visible keyboard focus indicator.

And then the details about contrasting area, etc go into the definition.

**Comment by mbgower** [comment]
2.4.7 doesn't get us much in the way of measurable normative language, but it at least provides a way to fail when something lacks a visible focus indicator.

Focus Appearance has 2 goals:
1. to define and give some muscle to what constitutes 'visible'
2. to ensure that the item receiving focus is itself visible, so that its focus indicator provides some benefit.

I used the "when" phrase to accomplish the second goal. Its purpose is to keep the item with focus from being obscured. Floating footers, non-modal dialogs, and a few other increasingly used authoring techniques have been making life harder for sighted keyboard users. This seemed like a solvable and worthy problem.

I had not thought about the "when" language being interpreted to allow or encourage fading indicators. It seems to me it should be easy to add some wording in the understanding document about persistence and intention.

I'm a little concerned with the perfect being the enemy of the good with some of our work. This SC in its current form will improve the experience of users. If it can be further improved upon, excellent. But the more we try to shore up against all possible forms of implementation or interpretation idiocy, the more unfathomable the wording threatens to become.

The current Focus Visible, with all its warts, has still had value. I think the same can be said of the new one, in its current form.

**Comment by kengdoj** [comment]
To respond to @bruce-usab asking if this detail is covered under Trusted Tester:

Trusted Tester's test condition for this is "A visible indication of focus is provided when focus is on the interface component." In order to pass, the focus must be visible while the component has focus. 

If a focus indicator were to fade to invisible while the component has focus, Trusted Tester would fail it for 2.4.7.

**Comment by patrickhlauke** [comment]
The Trusted Tester wording lacks the nuance of "has a mode of operation where the keyboard focus indicator is visible". How would TT work against things like `:focus-visible`, where you can have an outline/focus indicator when navigating with a keyboard, but suppress it when focus was received as a result of a mouse/touch interaction?

**Comment by kengdoj** [comment]
@patrickhlauke - sorry, I did not include the instructions for Trusted Tester in my earlier comment. Here is a fuller clip:

>Identify content to test: Use the keyboard to navigate to keyboard-accessible interface components (including drop-down menus, form fields, revealing/hiding content, tooltips, AND all interactive interface components). 

>How to test: Determine whether there is a visible indication of focus on the element that has keyboard focus.

>Test Condition (which the tester would pass/fail): A visible indication of focus is provided when focus is on the interface component.

The test would not cover when focus is received as a result of mouse/touch interaction.

**Comment by patrickhlauke** [comment]
@kengdoj ah, phew...that's good to know

**Comment by patrickhlauke** [comment]
as `:focus-visible` is evaluated/controlled by the browser, it allows for browsers to have options to always evaluate it to true regardless (rather than just using heuristics). in fact Chrome already has this (though not very well signposted) https://bugs.chromium.org/p/chromium/issues/detail?id=1272296

otherwise, the only answer to "'how does the designer know how any user is going to operate the revealed menu?" is "they can't know, so they must always assume keyboard", which is simply not going to fly as a requirement (*and* it would make for a fairly horrid user experience for mouse/touch users in many situations)

**Comment by JAWS-test** [comment]
@mbgower 

2.4.7 is now unfortunately formulated in such a way that it does not require a focus indicator for mouse operation, but for keyboard operation. But changing the operating mode (from mouse to keyboard) should not be a problem (not even in your example), because although I do not see any focus in mouse operation, the moment I start navigating with the keyboard (i.e. after the first arrow key operation) I see the focus. Thus 2.4.7 is fulfilled.

Of course, a focus criterion that also applies to mouse operation would be ideal, especially since 2.4.7 also mentions mouse users in Benefit.

**Comment by mbgower** [comment]
> the moment I start navigating with the keyboard... I see the focus

I can see some folks not agreeing with you that it has to be _navigation_. I think a very good case can be made that when I dismiss the dropdown with Esc, I am in a keyboard 'mode of operation' and the focus should be present.

> Of course, a focus criterion that also applies to mouse operation would be ideal, especially since 2.4.7 also mentions mouse users in Benefit.

I agree it should be beneficial. There are people who disagree with that, but I haven't seen implementations where clarity is lost or a user becomes confused because the keyboard focus is shown. If anyone has someone, I'd sure like to review and get a better understanding. The primary objections of designers tend to be it is inelegant (aka 'ugly', 'too strong', etc).

**Comment by patrickhlauke** [comment]
> The primary objections of designers tend to be it is inelegant (aka 'ugly', 'too strong', etc).

the secondary objection is that it makes "immediate action" type buttons all of a sudden look different - say you have a left and right arrow button on a carousel. you click with the mouse / tap on a touchscreen on one of them, now they look different. does a user understand that it just looks different because one now has focus, or is it because they now act/are different ... is one disabled one perhaps, is that why it has that difference in styling?

also, note that this has been browser behaviour for the default focus outline for ages ... browsers applying a "which input caused this to get focus" logic in deciding whether or not to show the default outline. so this should be tied to an option in the browsers.

> I can see some folks not agreeing with you that it has to be navigation. I think a very good case can be made that when I dismiss the dropdown with Esc, I am in a keyboard 'mode of operation' and the focus should be present.

then that argument needs to be made to browsers, because they're currently not doing that for their default focus outline nor the `:focus-visible` evaluation.

**Comment by patrickhlauke** [comment]
> The primary objections of designers tend to be it is inelegant (aka 'ugly', 'too strong', etc).

also, let's be clear... not just "designers", but actual clients/CEOs/etc that pay for sites to be made...

**Comment by mraccess77** [comment]
The keyboard user has this same issue "the secondary objection is that it makes "immediate action" type buttons all of a sudden look different - say you have a left and right arrow button on a carousel. you click with the mouse / tap on a touchscreen on one of them, now they look different. does a user understand that it just looks different because one now has focus, or is it because they now act/are different ... is one disabled one perhaps, is that why it has that difference in styling?"

So while that is an issue for touch/mouse users - we need to solve that in the future for keyboard only users as well.

**Comment by kengdoj** [comment]
@mbgower 
> Since a mouse opened the menu, no focus indicator needed. That seems like it is Trusted Testers stance.

Correct; Trusted Tester (TT) wouldn't check for a focus indicator here.

> I activate the dropdown by mouse, but I dismiss the dropdown menu by pressing Esc. So I've just used my keyboard. Should the focus now show on the dropdown target?

Hmmm, not by using the Esc key. The TT test instructions are "Use the keyboard to navigate". The Esc key doesn't shift focus so it would not be considered as keyboard navigation. The [keyboard navigation keys](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.keyboardnavigation?view=windowsdesktop-6.0#remarks) are Tab, Shift+Tab, Ctrl+Tab, Ctrl+Shift+Tab, Up Arrow, Down Arrow, Left Arrow, and Right Arrow. 

If, after clicking on the button with a mouse the next key press was a navigation key, a visible focus would be expected. For User X: select the GitHub Watch button by mouse, then press the up/down arrow key and the option with focus has a visible focus.

**Comment by patrickhlauke** [comment]
> The keyboard user has this same issue

arguably, not the same way, as the keyboard user is used to seeing an actual focus indication as they navigate through the page, so they're likely less surprised by the fact that the button that they've placed focus on now appears differently, and will likely infer from it that it's some form of focus indication.

**Comment by scottaohara** [comment]
>Correct; Trusted Tester (TT) wouldn't check for a focus indicator here.

that seems like a gap if so.  Especially in regards to test 4.H

**Comment by mraccess77** [comment]
"arguably, not the same way, as the keyboard user is used to seeing an actual focus indication as they navigate through the page, so they're likely less surprised by the fact that the button that they've placed focus on now appears differently, and will likely infer from it that it's some form of focus indication."

I'm not sure I agree - on many pages I find multiple types of indicators on the same page - sometimes 5 or more.  Also with a magnified view I may only find a few controls in view at a time - one button is white on green and the other is green on white - it's honestly not clear to a keyboard user which is the focused and which is the unfocused state.   Focus states are very often just inverted colors or similar and often don't look like indicators.  So it really depends on the quality and consistency of the indicator.

**Comment by mbgower** [comment]
This conversation has spun a bit out of orbit of the original question (and I'm one of the contributors to that). I just wanted to make sure that a [comment of mine from a week ago](https://github.com/w3c/wcag/issues/2219#issuecomment-1034355204) got some eye balls. I'm suggesting moving "when" out of the preamble into the second section. This is it in rough form:

>An area of the keyboard focus indicator meets the following:
Minimum area: The area is either:
at least as large as the area of a 1 [CSS pixel](https://w3c.github.io/wcag/guidelines/22/#dfn-css-pixels) thick [perimeter](https://w3c.github.io/wcag/guidelines/22/#dfn-perimeter) of the unfocused component, or
at least as large as the area of a 4 CSS pixel thick line along the shortest side of a [minimum bounding box](https://w3c.github.io/wcag/guidelines/22/#dfn-minimum-bounding-box) of the unfocused component, and no part of the area is thinner than 2 CSS pixels.
Contrast: The area has a contrast ratio of at least 3:1 between the colors when the component is focused and it is not focused.
Adjacent contrast: Where the area is adjacent to the component, it has a contrast ratio of at least 3:1 against the component or a thickness of at least 2 CSS pixels.
Additionally, when an item receives focus, it is not entirely hidden by author-created content.

**Comment by alastc** [comment]
The changes from #2281 separate the 'when' from the appearance criteria, so closing this one.
